ygnmi/ygnmi.go

```diff
@@ -303,6 +303,22 @@ func (w *Watcher[T]) Await() (*Value[T], error) {
 	return w.lastVal, err
 }
 
+// AwaitWithContext waits for the watch to finish in the same way as Await, but will
+// return immediately if the context supplied is completed.
+func (w *Watcher[T]) AwaitWithContext(ctx context.Context) (*Value[T], error) {
+	select {
+	case err, ok := <-w.errCh:
+		if !ok {
+			return nil, fmt.Errorf("Await already called and Watcher is closed")
+		}
+		close(w.errCh)
+		return w.lastVal, err
+	case <-ctx.Done():
+		close(w.errCh)
```

Comment:
I'm not sure if we should be closing the channel on the receiver side since the sending side might panic if it's concurrently trying to send a value.

Channels only need to be closed as a way of communicating end-of-data from the sender. My understanding is that existing code closes it because it *knows* the sender side won't send again; I think we should not have that assumption and simply return here without closing to avoid that happening.

To cancel the subscriber thread launched by `Watch` one option I can think of is to create a child context using `WithCancel` and then give the watcher object a reference to it so it can cancel that instead of closing the channel.

Reply:
ACK - agreed. I did not spend a lot of time to understand the design here. We can clearly remove the `close(w.errCh)` here.

I'm not quite sure which subscriber thread you're referring to here. It seems like we have a few different goroutines:

 * one that is where the `Await` is happening
 * one that is where the gNMI subscribe is happening
 * one where we're reading from the data and running the predicate

this change would make the `Await` goroutine listen to whether the context is done. As Daniel says below, using the same context here within the `Watch` call means that the gNMI subscribe would exit. I think we probably actually need to plumb this context into the one that is reading the data too (into the `select` at line 481). 

Which of these are you referring to when you say subscriber thread? Are there other (relevant) goroutines that are used here?

Reply:
I was originally thinking of just the predicate thread, so that the context reference knows when it can stop receiving values: https://github.com/openconfig/ygnmi/blob/830803303e619dc077a04e67f950f7d68b224838/ygnmi/ygnmi.go#L345C11-L345C14

I think it makes sense to cancel gNMI subscribe in the same way -- so it can tell the server to stop sending the data if Await no longer requires the updates.

Reply:
To clarify: what I mean is if the context passed to `Await` is done, then cancel the context passed to `Watch` (of course you would need to call `WithCancel` to get a cancel function).

Reply:
I think the user can do this themselves, right? In my calling code I am doing:

```
gnmiCtx, cancel := context.WithCancel(ctx)

watcher := ygnmi.WatchAll(gnmiCtx, c, ....)

if _, err := watcher.AwaitWithContext(gnmiCtx); err != nil { ... }
}
```

of course a user could decide to do both separately - if they wanted the granularity of being able to cancel `Await` and `WatchAll` separtely.

Reply:
Resolved, see note for future code historians.